### PR TITLE
UEFI Boot support

### PR DESCRIPTION
**Pull request by mintsoft**
Hiya,

Based on the work by @cab404 : https://github.com/Tomas-M/linux-live/pull/75 with bits of the receipe from https://github.com/Tomas-M/linux-live/pull/75#issuecomment-535481117  by @felixlohmeier 

I've moved all the file "logic" into the ./build so that the Slax and the main `bootinst.sh/.bat` can be simple.

I've just built this end-to-end and tested on a ProxMox VM + Dell XPS 13 in UEFI booting and it works a charm. This is using the non-vesa `menu.c32`. I'm still working on getting that working, however in the meantime this is fully functional from my testing.

### CODE REVIEW

**Comment by mintsoft** [comment]
@Tomas-M additionally I *think* I have got vesamenu working (https://github.com/mintsoft/linux-live/tree/uefi-vesamenu), but it requires `6.04-pre1` which is a pre-release version; I can update this PR with that one if desired.

**Comment by felixlohmeier** [comment]
I have just tested the pull request with Lubuntu 18.04.

* Create linux-live-kit zip archive
```
sudo apt install aufs-tools squashfs-tools
cd /tmp
git clone https://github.com/mintsoft/linux-live.git --branch uefi --single-branch
cd linux-live
sudo ./build
sudo /tmp/gen_linux_zip.sh
```
* Extract `/tmp/linux-x86_64.zip` to usb drive
* move folder `linux-x86_64/linux` into root directory of usb drive
```
mv
```
* remount usb drive with full permissions and execute `bootinst.sh`
```
sudo umount /dev/sda1
mkdir usb
sudo mount /dev/sda1 usb
cd usb/linux/boot
sudo ./bootinst.sh
```

Results:
* [Line 53 in build.sh](https://github.com/Tomas-M/linux-live/pull/84/files#diff-b0da275520918e23dd615e2a747528f1R53) now produces a warning because cp is (intentionally) used without -r and the directory `bootfiles` contains a subdirectory `efi-files`. Error message (with German locals):
```
bain@bain-VirtualBox:/tmp/linux-live$ sudo ./build
Changing current directory to /tmp/linux-live
Building intramfs image...
cp: -r wurde nicht angegeben, daher wird das Verzeichnis 'bootfiles/efi-files' ausgelassen
Parallel mksquashfs: Using 2 processors
```
* [Lines 43 and 44 in bootfiles/bootinst.sh](https://github.com/Tomas-M/linux-live/pull/84/files#diff-c2be99450d0099233c75d62f87b1afc4R43) do not work as expected. Files should be copied to the root directory of the usb drive. Instead they will be copied to `linux/boot/EFI/Boot`.
```
[felix@tux boot]$ sudo ./bootinst.sh
. is device /dev/sda1
cp: cannot stat 'boot/EFI/Boot/*': No such file or directory
Boot installation finished.
```

File: bootfiles/bootinst.sh [hunk 1]
```diff
@@ -40,5 +40,9 @@ if [ "$DEV" != "$PART" ]; then
    ) | fdisk $DEV >/dev/null 2>&1
 fi
 
+# UEFI boot loader
```
Comment (felixlohmeier) [inline bootfiles/bootinst.sh]: Test result in https://github.com/Tomas-M/linux-live/pull/84#issuecomment-537543083:
> Files should be copied to the root directory of the usb drive. Instead they will be copied to `linux/boot/EFI/Boot`.

Proposed solution:
```
# UEFI boot loader
mkdir -p "$BOOT/../../EFI"
mv "EFI/Boot" "$BOOT/../../EFI/"
```

Comment (mintsoft) [reply bootfiles/bootinst.sh]: Implemented this, thanks for testing!

**Comment by mintsoft** [comment]
Awesome, thanks for testing that @felixlohmeier I've fixed up the `bootinst.sh`.

> 
>    Line 53 in build.sh now produces a warning because cp is (intentionally) used without -r and the directory bootfiles contains a subdirectory efi-files. Error message (with German locals):

>bain@bain-VirtualBox:/tmp/linux-live$ sudo ./build
>Changing current directory to /tmp/linux-live
>Building intramfs image...
>cp: -r wurde nicht angegeben, daher wird das Verzeichnis 'bootfiles/efi-files' ausgelassen
>Parallel mksquashfs: Using 2 processors

I'm not so sure what to do about that one; I'm tempted to just make it recursively copy as it won't do any harm per-se, just seems like a waste of space

**Comment by felixlohmeier** [comment]
Suppressing the warning is really not very important, but I would have a suggestion that might work. The folder name `efi-files` isn't used anywhere else, is it? What about moving files in git repo from `bootfiles/efi-files` to `bootfiles/EFI/Boot`?
* [Line 53 in build](https://github.com/Tomas-M/linux-live/pull/84/files#diff-b0da275520918e23dd615e2a747528f1R53) `cp bootfiles/* $BOOT` with an additional `-r` will then copy boot and uefi files into the expected folder structure in one step.
* Line 59 (`mkdir -p $BOOT/EFI/Boot`) and line 61 (`cp bootfiles/efi-files/{ldlinux.e64,menu.c32,libutil.c32} $BOOT/EFI/Boot`) could be deleted
* Line 60 could be changed to a mv command, e.g. `mv $BOOT/EFI/Boot/syslinux.efi $BOOT/EFI/Boot/bootx64.efi`

I hope this works, it's not tested.

**Comment by mintsoft** [comment]
@felixlohmeier I've just done that; actually simplifies things significantly

**Comment by felixlohmeier** [comment]
I have successfully tested the updated pull request:

* Create linux-live-kit zip archive on Lubuntu 18.04
  ```
  sudo apt install aufs-tools squashfs-tools
  cd /tmp
  git clone https://github.com/mintsoft/linux-live.git --branch uefi --single-branch
  cd linux-live
  sudo ./build
  sudo /tmp/gen_linux_zip.sh
  ```
* Create bootable uefi usb drive (MBR/FAT32) from generated zip
  * Linux (Fedora29)
    * Extract `/tmp/linux-x86_64.zip` to usb drive
    * move folder `linux-x86_64/linux` into root directory of usb drive
    * remount usb drive with full permissions and execute `bootinst.sh`
    ```
    sudo umount /dev/sda1
    mkdir usb
    sudo mount /dev/sda1 usb
    cd usb/linux/boot
    sudo ./bootinst.sh
    ```
  * Windows 10: 
    * Extract `/tmp/linux-x86_64.zip` to usb drive
    * move folder `linux-x86_64/linux` into root directory of usb drive
    * Start PowerShell as administrator, move to directory linux/boot on usb drive, e.g. `cd E:\linux\boot` and execute script with `.\bootinst.bat`

@tomas-m: Successfully tested on Lubuntu 18.04 with bootinst.sh (on Fedora29) and bootinst.bat (on Windows 10). I suggest to merge this pull request.

**Caveat:** UEFI boot will only work on machines with Secure Boot disabled. Syslinux does not support Secure Boot.

**Comment by felixlohmeier** [comment]
> Caveat: UEFI boot will only work on machines with Secure Boot disabled. Syslinux does not support Secure Boot.

Maybe we should add a note in https://github.com/Tomas-M/linux-live/blob/master/bootinfo.txt? The following instructions should be helpful: http://www.rodsbooks.com/efi-bootloaders/secureboot.html#disable

**Comment by ryans-git** [comment]
Hi @felixlohmeier is there a way to get a UEFI iso for a DVD? Can't seem to get anything to boot into EFI (only legacy so I can't use and efi shell commands in the live dvd). I've tried variations of:
```
cd /tmp/linux-data-2022 && \
 xorriso -as mkisofs -o /tmp/linux-live-name.iso -v -J -A linux  \
-isohybrid-mbr linux/boot/mbr.bin -c linux/boot/isolinux.boot -b linux/boot/isolinux.bin  \
-no-emul-boot -boot-info-table -boot-load-size 4 \
-e linux/boot/EFI/Boot/bootx64.efi -isohybrid-gpt-basdat .
```
But can't seem to get it to work. Looking at `bootinst.sh`, do I need to run `extlinuix.x86_64 --install` or move the EFI/boot directory around in the root directory before the mkisofs? 

The bootable usb on CentOS and Windows 10 instructions work great btw, just trying to get an iso for a DVD now.

**Comment by mintsoft** [comment]
@ryans-git I've not tried that yet, but the ISO produced will basically need to have `/EFI/Boot/bootx64.efi` in its root (this is the EFI boot program that's used). Does the `-e` argument to xorriso do that?

**Comment by mintsoft** [comment]
@Tomas-M I can see that you're starting to add Debian 9/10 support. Do you want this PR to work for both or is there no point in supporting Debian 9 anymore?

**Comment by ryans-git** [comment]
@mintsoft It works using your suggestion to move `EFI/Boot/Bootx64.efi` to the root.

```
cd /tmp/linux-data-XX
mv linux/boot/EFI .
xorriso -as mkisofs \
        -o /tmp/uefi-dvd-name.iso -v -J -A linux \
        -isohybrid-mbr linux/boot/mbr.bin -b linux/boot/isolinux.bin \
        -no-emul-boot -boot-info-table -boot-load-size 4 \
        -e EFI/Boot/bootx64.efi \
        .
```

I mostly just did `-e` because of other examples I've seen. From the man page I'm still not sure which option I should be using to point to the efi image.

> **-e** _iso_rr_path_
>     Specify the boot image file which shall be mentioned in the current entry of the El Torito boot 
>     catalog. It will be marked as suitable for EFI.
>     Normally no other El Torito options should be used with the catalog entry that points to an EFI 
>     image. Consider to use --efi-boot rather than -e.
> **--efi-boot** _iso_rr_path_
>     Perform -eltorito-alt-boot, option -e with the given iso_rr_path, and again -eltorito-alt-boot. This 
>     gesture is used for achieving EFI-bootability of the GRUB2 rescue CD.

Anyway, the command works. Thanks a lot for the help.

**Comment by LothaireCliquennois** [comment]
What will happen with this PR ?

**Comment by mintsoft** [comment]
Good question, I can resolve the merge conflicts and make it support both the new Debian versions but I'm not planning to do so unless it's actually going to get merged/used at some point

**Comment by Tomas-M** [comment]
First of all, thank you very much for such effort in implementing and testing UEFI booting for linux-live (and Slax). I hate the fact that it must use /EFI/boot directory on the USB drive, but that is not your fault, that is the fault of the idiot who developed EFI standard. Your solution to create the /EFI/boot directory in bootinst script is acceptable.

When I reviewed it 7 months ago here on github, I told you that I had an objection against newlines in bootinst.bat. I was waiting for you to fix that, but it is possible that you missed my comment, or you do not understand the reasoning.

It is following: bootinst.bat should work on Linux too. For that reason, first 3 lines of bootinst.bat must use Linux line ending (LF), while all the other lines must use Windows line endings (CRLF). This hack makes the same script (bootinst.bat) usable on both Windows and Linux, where on Linux it simply calls bootinst.sh. This is useful when user has USB drive mounted with showexec mount option by default (as is used on some Linuxes). If set, the execute permission bits of the file will be allowed only if the extension part of the name is .EXE, .COM, or .BAT, so user could not run bootinst.sh easily.

Anyway, I can fix that myself later.

I have few more questions

- I can see you are using ldlinux.e64, do I assume correctly that the whole UEFI boot will only work on 64bit systems? Is there no support for ldlinux.e32?

- you mentioned vesa menu is working only with syslinux-6.04-pre release. Is this version used only for the UEFI stuff? If so then I don't mind.

Thanks

**Comment by Tomas-M** [comment]
One more thing to add, I would like support for both Debian9 and Debian10.

**Comment by mintsoft** [comment]
Cool; I can easily fix the line endings and remerge in for D9/D10.  

I have no idea where the comment you've put on the PR is @Tomas-M as I can't see it anywhere :)

>I can see you are using ldlinux.e64, do I assume correctly that the whole UEFI boot will only work on 64bit systems? Is there no support for ldlinux.e32?

As it stands yeah this will only work on 64Bit EFI; I don't have a 32bit EFI machine to test it on (They're pretty rare, the only device I know of is the 1st generation MacMini).

>you mentioned vesa menu is working only with syslinux-6.04-pre release. Is this version used only for the UEFI stuff? If so then I don't mind.

Again yes, that's only the EFI bootloader; I've not touched the BIOS syslinux at all.

**Comment by mintsoft** [comment]
@Tomas-M I've just reworked this so that it supports both D9 and D10. I've ensured that the line endings are preserved in the `.bat`` files and it seems to be working for me

@felixlohmeier Is there any chance you could give this another runaround? It works for me but you found a few things I missed last time :)

**Comment by felixlohmeier** [comment]
Successfully tested on Debian 10 with bootinst.sh (on Fedora 31) and bootinst.bat (on Windows 10). Similar test procedure as in https://github.com/Tomas-M/linux-live/pull/84#issuecomment-538346820.

**Comment by Tomas-M** [comment]
Thank you

**Pull request merged**